controllers/builder.go

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package controllers
+
+import (
+	"fmt"
+	"github.com/davidesalerno/kogito-serverless-operator/constants"
+	"github.com/ricardozanini/kogito-builder/api"
+	"github.com/ricardozanini/kogito-builder/builder"
+	"github.com/ricardozanini/kogito-builder/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"time"
+)
+
+func Build(sourceSwf []byte, dockerFile []byte) (bool, error) {
+	cli, err := client.NewClient(true)
+	platform := api.PlatformBuild{
+		ObjectReference: api.ObjectReference{
+			Namespace: constants.BUILDER_NAMESPACE_DEFAULT,
+			Name:      constants.BUILDER_IMG_NAME_DEAULT,
+		},
+		Spec: api.PlatformBuildSpec{
+			BuildStrategy:   api.BuildStrategyPod,
+			PublishStrategy: api.PlatformBuildPublishStrategyKaniko,
+			Registry: api.RegistrySpec{
+				Insecure: true,
+			},
+			Timeout: &metav1.Duration{
+				Duration: 5 * time.Minute,
+			},
+		},
+	}
+
+	build, err := builder.NewBuild(platform, "greetings:latest", "kogito-test").
+		WithResource("Dockerfile", dockerFile).WithResource("greetings.sw.json", sourceSwf).
+		WithClient(cli).
+		Schedule()
+	if err != nil {
+		fmt.Println(err.Error())
+		return false, err
+		//panic("Can't create build")
+	}
+
+	// from now the Reconcile method can be called until the build is finished
+	for build.Status.Phase != api.BuildPhaseSucceeded &&
+		build.Status.Phase != api.BuildPhaseError &&
+		build.Status.Phase != api.BuildPhaseFailed {
+		fmt.Printf("\nBuild status is %s", build.Status.Phase)
+		build, err = builder.FromBuild(build).WithClient(cli).Reconcile()
+		if err != nil {
+			fmt.Println("Failed to run test")
+			//panic(fmt.Errorf("build %v just failed", build))
+			return false, err
+		}
+		time.Sleep(10 * time.Second)
```

Comment:
Don't do this here. The example I left in the PoC was to mimic the controller reconciliation pattern. So you trigger a reconciliation here, each build state then will be triggered by the Kubernetes APIs, not sleep.

Reply:
@ricardozanini  Ok, but do we want a fire and forget behaviour instead to return the result of the build ?

Reply:
Take a look at the Camel-K controller/builder package. You will see how they do.

Basically every time k8s does a reconciliation, this method will be called, and you advance the action automatically. You don't need this loop here.

Reply:
Ok